### PR TITLE
📖  Update the Kubebuilder book tutorials for v3

### LINKED ISSUE #2273: Code generated by kubebuilder init doesn't match code in book
### What broke? What's expected?

I'm running into compilation problems trying to follow along in the book. Specifically, the [Implementing a controller](https://book.kubebuilder.io/cronjob-tutorial/controller-implementation.html) chapter.

The problem seems to be that the code generated by `kubebuilder init` is different from the code in the book. A few examples:

* The kubebuilder command generates the `batchv1 "tutorial.kubebuilder.io/project/api/v1"` import, but the book shows a `batch "tutorial.kubebuilder.io/project/api/v1"` import in _controllers/cronjob_controller.go_ (`batch` vs. `batchv1`).
* The book shows getting a `log` object from the `CronJobReconciler` struct, but the generated code instead uses `_ := log.FromContext(ctx)`. It's unclear to me which way of assigning `log` is correct, or how to resolve this inconsistency.

There may be other issues, but I decided to stop here and file an issue once I realized that this might be a doc issue.

As someone relatively new to Go, it took me a while to figure out how to work around these inconsistencies. My expectation was that the code in the book would match the code generated by `kubebuilder init`, and that I could copy/paste the code into my project as I went along.

### Reproducing this issue

Follow the steps in [Tutorial: Building CronJob](https://book.kubebuilder.io/cronjob-tutorial/cronjob-tutorial.html).

See below for the versions of the tools that I'm using.

### KubeBuilder (CLI) Version

Version: main.version{KubeBuilderVersion:"3.1.0", KubernetesVendor:"1.19.2", GitCommit:"92e0349ca7334a0a8e5e499da4fb077eb524e94a", BuildDate:"2021-05-27T17:54:28Z", GoOs:"linux", GoArch:"amd64"}

### PROJECT version

3

### Plugin versions

```yaml
- go.kubebuilder.io/v3
```

### Other versions

* go version go1.15.7 linux/amd64
* sigs.k8s.io/controller-runtime v0.8.3
* (no controller-tools in `go.mod` file)

### Extra Labels

/kind documentation

### PR DESCRIPTION
**Pull request by hickeyma**
This PR updates the Kubebuilder book tutorials to align with the current version, v3. This fix will help users navigate the book with success.

Fixes #2273

### CODE REVIEW

**Comment by hickeyma** [comment]
This is a WIP as I traverse the book. The initial docs (intro, quick start and arch) are reviewed. Next will be the cronjob tutorial.

**Comment by hickeyma** [comment]
@camilamacedo86 First commit after running  https://github.com/kubernetes-sigs/kubebuilder/blob/master/docs/book/src/cronjob-tutorial/testdata/generate_cronjob.sh. Let me kjnow if going in the right direction.

File: docs/book/src/cronjob-tutorial/testdata/project/api/v1/cronjob_types.go [hunk 1]
```diff
@@ -13,160 +13,36 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
-	// +optional
-	LastScheduleTime *metav1.Time `json:"lastScheduleTime,omitempty"`
 }
 
-/*
-Finally, we have the rest of the boilerplate that we've already discussed.
-As previously noted, we don't need to change this, except to mark that
-we want a status subresource, so that we behave like built-in kubernetes types.
-*/
-
```
Comment (camilamacedo86) [inline docs/book/src/cronjob-tutorial/testdata/project/api/v1/cronjob_types.go:162]: You need to come back to the texts on top as well. 
I usually compare that with the IDEA and ensure that we just update what changed. 
So, you need to 
a) run the scripts to re-gen the scaffolds ( base ones )
b) run make in the dirs of the projects to generate the files
c) revert the files that was removed and are missing 
d) compare with master to add again the code on top + text and explanations.

Comment (hickeyma) [reply docs/book/src/cronjob-tutorial/testdata/project/api/v1/cronjob_types.go:162]: Thanks for the feedback @camilamacedo86. I have updated the commit. Some build issues to cleanup.

**Comment by hickeyma** [comment]
Unit test failing. Just wondering if this is a glitch or not?

```
STEP: tearing down the test environment
Failure [20.007 seconds]
[AfterSuite] AfterSuite 
/home/runner/work/kubebuilder/kubebuilder/docs/book/src/cronjob-tutorial/testdata/project/controllers/suite_test.go:152

  Unexpected error:
      <errors.aggregate | len:1, cap:1>: [
          <*errors.errorString | 0xc0008499d0>{
              s: "timeout waiting for process kube-apiserver to stop",
          },
      ]
      timeout waiting for process kube-apiserver to stop
  occurred

  /home/runner/work/kubebuilder/kubebuilder/docs/book/src/cronjob-tutorial/testdata/project/controllers/suite_test.go:155
```

**Comment by camilamacedo86** [comment]
HI @hickeyma, 

We need to solve : https://github.com/kubernetes-sigs/controller-runtime/issues/1571

File: docs/book/src/cronjob-tutorial/testdata/project/controllers/suite_test.go [hunk 2]
```diff
@@ -127,22 +126,24 @@ var _ = BeforeSuite(func() {
 		available when talking directly to the API server.
 	*/
 
-	k8sManager, err := ctrl.NewManager(cfg, ctrl.Options{
+	//TODO: Fix this as it is disabled because following error when tearing down:
```
Comment (camilamacedo86) [inline docs/book/src/cronjob-tutorial/testdata/project/controllers/suite_test.go:129]: If we add the TODO here then that will appears the docs :-(

**Comment by hickeyma** [comment]
Thanks for the feedback @camilamacedo86. Blocked waiting on https://github.com/kubernetes-sigs/kubebuilder/pull/2325

File: docs/book/src/cronjob-tutorial/testdata/project/controllers/suite_test.go (hunk 2 above)
Comment (hickeyma) [reply docs/book/src/cronjob-tutorial/testdata/project/controllers/suite_test.go]: Test is enabled again

**Comment by hickeyma** [comment]
@camilamacedo86 I updated the failing tests and they are now passing. I have updated the controller-runtime `envTest` issue in https://github.com/kubernetes-sigs/controller-runtime/issues/1571#issuecomment-914401589. I had to change how the test environment is brought down when a controller is started. I have updated `cronjob_tutoria`l test in this PR accordingly.

**Review by estroz — APPROVED** [review]
The tutorial envtest bits will need to be changed when teardown starts working properly again. For now the comments are explanatory enough.

/lgtm
/approve

**Comment by hickeyma** [comment]
Thanks for the reviews @camilamacedo86 @estroz, and for merging @estroz.

**Comment by estroz** [comment]
Thanks for your contribution!

**Pull request merged**